cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,491 @@
+
+    def _make_numexpr(self, cpm_expr):
+        """
+            Converts a numeric CPMpy 'flat' expression into a solver-specific numeric expression
+
+            Primarily used for setting objective functions, and optionally in constraint posting
+        """
+
+        # [GUIDELINE] not all solver interfaces have a native "numerical expression" object.
+        #       in that case, this function may be removed and a case-by-case analysis of the numerical expression
+        #           used in the constraint at hand is required in __add__
+        #       For an example of such solver interface, check out solvers/choco.py or solvers/exact.py
+
+        if is_num(cpm_expr):
+            return cpm_expr
+
+        # decision variables, check in varmap
+        if isinstance(cpm_expr, _NumVarImpl):  # _BoolVarImpl is subclass of _NumVarImpl
+            return self.solver_var(cpm_expr)
+
+        # any solver-native numerical expression
+        if isinstance(cpm_expr, Operator):
+           if cpm_expr.name == 'sum':
+               return self.TPL_solver.sum(self.solver_vars(cpm_expr.args))
+           elif cpm_expr.name == 'wsum':
+               weights, vars = cpm_expr.args
+               return self.TPL_solver.weighted_sum(weights, self.solver_vars(vars))
+           # [GUIDELINE] or more fancy ones such as max
+           #        be aware this is not the Maximum CONSTRAINT, but rather the Maximum NUMERICAL EXPRESSION
+           elif cpm_expr.name == "max":
+               return self.TPL_solver.maximum_of_vars(self.solver_vars(cpm_expr.args))
+           # ...
+        raise NotImplementedError("TEMPLATE: Not a known supported numexpr {}".format(cpm_expr))
+
+
+    # `__add__()` first calls `transform()`
+    def transform(self, cpm_expr):
+        """
+            Transform arbitrary CPMpy expressions to constraints the solver supports
+
+            Implemented through chaining multiple solver-independent **transformation functions** from
+            the `cpmpy/transformations/` directory.
+
+            See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations
+        cpm_cons = toplevel_list(cpm_expr)
+        # count is only supported with a constant to be counted, so we decompose
+        supported = {"alldifferent", 'inverse', 'nvalue', 'element', 'table', 'indomain',
+                     "negative_table", "gcc", 'max', 'min', 'abs'}
+        supported_reified = {"alldifferent", 'nvalue', 'element', 'table', 'indomain', 'max', 'min',
+                     "negative_table", 'abs'}
+        cpm_cons = decompose_in_tree(cpm_cons, supported=supported, supported_reified=supported_reified)
+        '''cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
```

Comment:
so, without flattening? you can remove this then, maybe add a comment line that it indeed does not require flattening...

Reply:
Ah yes, still some comments that i have to clean up it seems